### PR TITLE
Encode plus to %2B for S3 copies

### LINKED ISSUE #88: Problems handling S3 keys with spaces and/or encoded spaces
We're seeing some issues with the `CopyObject` method on the Javascript SDK handling source files that include either spaces or `+` in the object key (unclear which, maybe both). Encoding the source object doesn't seem to be doing the trick

### PR DESCRIPTION
**Pull request by farski**
When passing an object to the CopySource of copyObject, if the object key includes a plus character, the SDK/API will only properly handle that if it is encoded to %2B.

The S3 API/SDK expects this value to be URL encoded in a way that preserves forward slashes (among other things). Normally that sort of encoding (URL vs component) would not encoded pluses. Since it's necessary here, the encoding must be done manually, *after* the rest of the string has been URL encoded as usual.

- [x] Fix encoding issue
- [x] Write some tests

Closes #88

### CODE REVIEW

File: test/acceptance/copy_test.rb [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+    describe :to_key_with_spaces do
+      it 'handles HTTP source files with spaces' do
+        req = step_functions.start_execution({
+          state_machine_arn: CONFIG.PORTER_STATE_MACHINE_ARN,
+          input: {
+            Job: {
+              Id: 'porter-test-copy-http-spaces-to-spaces',
+              Source: {
+                Mode: 'HTTP',
+                URL: 'https://dovetail-cdn-staging.prxu.org/815/dAsUFiIz68rzpns5Cv7XM36Pt8bBzbu7gn3ui6Hg8-o/input file with spaces.mp3'
```
Comment (cavis) [inline test/acceptance/copy_test.rb:20]: I mean, technically this is an invalid URL.

File: test/acceptance/copy_test.rb [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+              Source: {
+                Mode: 'HTTP',
+                URL: 'https://dovetail-cdn-staging.prxu.org/815/dAsUFiIz68rzpns5Cv7XM36Pt8bBzbu7gn3ui6Hg8-o/input+file+with+spaces.mp3'
+              },
+              Tasks: [
+                {
+                  Type: "Copy",
+                  Mode: "AWS/S3",
+                  BucketName: CONFIG.PORTER_TEST_BUCKET_NAME,
+                  ObjectKey: "copied file with spaces"
```
Comment (cavis) [inline test/acceptance/copy_test.rb:125]: I would expect this to be `copied+file+with+spaces` (literal plus signs).

Comment (cavis) [reply test/acceptance/copy_test.rb:125]: Oh, I see ... this is the caller specifying they want to change the filename.  Nevermind!

**Review by cavis — COMMENTED** [review]
Questions!  And some of those acceptance tests seem redundant ... but I suppose it doesn't hurt anything.

File: lambdas/CopyLambdaFunction/index.js [hunk 3]
```diff
@@ -34,7 +34,7 @@ async function awsS3copyObject(event) {
   const params = {
     CopySource: encodeURI(
       `/${event.Artifact.BucketName}/${event.Artifact.ObjectKey}`,
-    ),
+    ).replace(/\+/g, '%2B'),
```
[the remark this answers is not in the document]
Comment (farski) [reply lambdas/CopyLambdaFunction/index.js:37]: > Bucket [names](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html) can consist only of lowercase letters, numbers, dots (.), and hyphens (-).

File: test/acceptance/copy_test.rb (hunk 1 above)
Comment (farski) [reply test/acceptance/copy_test.rb:20]: Do you think Porter should reject it?

Comment (farski) [reply test/acceptance/copy_test.rb:20]: If we don't, worst case is it fails, best case it gets encoded to what they actually wanted.

Comment (cavis) [reply test/acceptance/copy_test.rb:20]: Nah- I think attempting to handle it is better than nothing.  Which is what you're doing.  Resolved!

File: lambdas/IngestLambdaFunction/index.js [hunk 4]
```diff
@@ -85,7 +85,9 @@ function httpGet(uri, file, redirectCount) {
 function filenameFromSource(source) {
   if (source.Mode === 'HTTP') {
     const urlObj = url.parse(source.URL);
-    return urlObj.pathname.split('/').pop() || urlObj.hostname;
+    return (
+      decodeURIComponent(urlObj.pathname.split('/').pop()) || urlObj.hostname
```
[the remark this answers is not in the document]
Comment (farski) [reply lambdas/IngestLambdaFunction/index.js:89]: If that throws, the Lambda will fail and the Step Function execution will return a `SOURCE_FILE_INGEST_ERROR` state, which I think is the best we can do with a funky URL. Can you think of anything we could do if we catch?

[the remark this answers is not in the document]
Comment (cavis) [reply lambdas/IngestLambdaFunction/index.js:89]: Nah, just asking making sure you weren't expecting this to "never ever throw".

**Review by cavis — APPROVED** [review]
👍

**Pull request merged**